src/Avalonia.Controls.TreeDataGrid/Selection/TreeDataGridRowSelectionModel.cs

```diff
@@ -184,6 +188,85 @@ private void Search(TreeDataGrid treeDataGrid, string candidatePattern, int sele
+                        {
+                            if (childrenCount + _lastPageSelectedIndex <= sender.RowsPresenter.Items.Count)
+                            {
+                                _lastPageSelectedIndex = childrenCount - 1 + _lastPageSelectedIndex;
+                            }
+                            else
+                            {
+                                _lastPageSelectedIndex = sender.RowsPresenter.Items.Count - 1;
+                            }
+
+                        }
+                        else if (_lastPageSelectedIndex == 0 || SelectedIndex != _lastPageSelectedIndex)
+                        {
+                            var max = -1;
+                            for (int i = 0; i <= childrenCount - 1; i++)
+                            {
+                                if (children.ElementAt(i) is TreeDataGridRow row &&
+                                    row.RowIndex > max &&
+                                    row.TransformedBounds != null &&
+                                    IsElementFullyVisibleToUser(row.TransformedBounds.Value))
+                                {
+                                    max = row.RowIndex;
+                                }
+                            }
+                            _lastPageSelectedIndex = max;
+
+                        }
+                    }
+                    else if (e.Key == Key.PageUp)
+                    {
+                        if (_lastPageSelectedIndex > 0 && SelectedIndex == _lastPageSelectedIndex)
+                        {
+                            if (_lastPageSelectedIndex - childrenCount >= 0)
+                            {
+                                _lastPageSelectedIndex = _lastPageSelectedIndex - childrenCount + 1;
+                            }
+                            else
+                            {
+                                _lastPageSelectedIndex = 0;
+                            }
+
+                        }
+                        else if (_lastPageSelectedIndex == 0 || SelectedIndex != _lastPageSelectedIndex)
+                        {
+                            var min = -1;
+                            for (int i = 0; i <= childrenCount - 1; i++)
+                            {
+                                if (children.ElementAt(i) is TreeDataGridRow row &&
+                                    ((row.RowIndex < min &&
+                                    row.TransformedBounds != null &&
+                                    IsElementFullyVisibleToUser(row.TransformedBounds.Value)) || min == -1))
+                                {
+                                    min = row.RowIndex;
+                                }
+                            }
+                            _lastPageSelectedIndex = min;
+                        }
+                    }
```

Comment:
In general, I wonder if this logic couldn't be simplified - a lot of it looks the same except that it's searching in a different direction depending on whether page up or down was pressed.